### PR TITLE
#1 create gcn2hop module

### LINKED ISSUE #1: Productize gcn2scimma code
- [x] convert to using the cookiecutter template
- [ ] add unit tests
- [ ] add continuous integration
- [x] add linting, make sure flake8 checks pass
- [x] convert `gcn2scimma.py` to module+entrypoint style, like we have in scimma-client. This will avoid the problem of which interpreter path to hardcode at the top of `gcn2scimma.py`

### PR DESCRIPTION
**Pull request by shereenElSayed**
- I changed the code where you the old flow will still be working along with the module+entrypoint. 

- Enhancement and/or better code in the makefile is possible. I added new targets for linting and formatting

- I added setup.py for installation using `python setup.py install`, setup.cfg for flake8 configurations

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -2,28 +2,53 @@
...
+install:
 	mkdir -p $(INSTALL_DIR)
-
-$(INSTALL_DIR)/%: %
-	cp $< $@
-	chmod ugo+rx $@
-
-install: install_dir $(INSTALL_FILES)
+	cp $(WORKING_DIR)/{$(PYTHON_FILES)} $(INSTALL_DIR)
+	chmod ugo+rx $(WORKING_DIR)/{$(PYTHON_FILES)}
```
Comment (myNameIsPatrick) [inline Makefile]: By moving to `setup.py`, you should only need to run `$PYTHON setup.py install` to install the module in a place that's discoverable by your environment.

File: Makefile [hunk 2]
```diff
@@ -2,28 +2,53 @@
 ## Install gcn2scimma
 ##
 INSTALL_DIR   := /usr/local/gcn2hop
-PYTHON_FILES  := gcn2hop Utils.py 
+WORKING_DIR	  := ./gcn2hop
+PYTHON_FILES  := gcn2hop,Utils.py
 INSTALL_FILES := $(patsubst %,$(INSTALL_DIR)/%,$(PYTHON_FILES))
```
Comment (myNameIsPatrick) [inline Makefile]: This shouldn't be necessary either after moving to a module.

File: setup.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    "pygcn"
+]
+extras_require = {
+    'dev': ['pytest', 'pytest-console-scripts', 'pytest-cov', 'flake8', 'flake8-black'],
+}
+
+setup(
+    name = 'gcn2hop',
+    url = 'https://github.com/scimma/gcn2hop',
+    author = 'Ron Tabia',
```
Comment (myNameIsPatrick) [inline setup.py]: Change author to `Ron Tapia`.

File: setup.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        'Intended Audience :: Science/Research',
+        'Intended Audience :: Developers',
+        'Natural Language :: English',
+        'Topic :: Scientific/Engineering',
+        'Topic :: Scientific/Engineering :: Astronomy',
+        'Topic :: Scientific/Engineering :: Physics',
+        'Operating System :: POSIX',
+        'Operating System :: Unix',
+        'Operating System :: MacOS',
+        'License :: OSI Approved :: GNU General Public License',
```
Comment (myNameIsPatrick) [inline setup.py]: Change to `License :: OSI Approved :: GNU General Public License v3 or later (GPLv3+)`

File: setup.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+]
+extras_require = {
+    'dev': ['pytest', 'pytest-console-scripts', 'pytest-cov', 'flake8', 'flake8-black'],
+}
+
+setup(
+    name = 'gcn2hop',
+    url = 'https://github.com/scimma/gcn2hop',
+    author = 'Ron Tabia',
+    license = 'GNU GENERAL PUBLIC',
```
Comment (myNameIsPatrick) [inline setup.py]: Change license to `GPLv3+`.

File: tests/test_basic.py [hunk 6]
```diff
@@ -1,6 +1,7 @@
 import os
 
-def test_basic ():
+
+def test_basic():
     cmd = "/usr/local/gcn2hop/gcn2hop --help"
```
Comment (myNameIsPatrick) [inline tests/test_basic.py]: After the install target in the Makefile is changed appropriately, this `cmd` should work with `gcn2hop --help`.

File: Makefile [hunk 7]
```diff
@@ -2,28 +2,53 @@
...
+.PHONY: help
+help :
+	@echo
+	@echo 'Commands:'
+	@echo
+	@echo '  make test                  run unit tests'
+	@echo '  make lint                  run linter'
+	@echo '  make format                run code formatter, giving a diff for recommended changes'
+	@echo '  make doc                   make documentation'
+	@echo '  make install				install module
```
Comment (myNameIsPatrick) [inline Makefile]: Please fix the indentation stuff for the help messages here.

**Review by myNameIsPatrick — CHANGES_REQUESTED** [review]
Hi @shereenElSayed, I added my review with some requested changes, focused purely on making this library discoverable after installing with `setup.py`.

The types of changes fall mainly into two classes; (1) fixing typos in `setup.py` with licensing and such, and (2) allowing the installation to leverage `setup.py` and having the tests reflect that.

I haven't tested the changes myself yet, but I'll run through them locally after the request changes are put in. Overall, looking good!

Out of scope for this PR but I think would be good to keep on the radar are:
  * Remove the use of globals in favor of partial function evaluation (`functools.partial`) to allow `gcn.voevent.listen` to work correctly with the stream connection we're passing in.
  * Move from optparse to argparse. optparse is deprecated for python 3, although still working. It's not clear when/if this behavior will break, but given that the python language doesn't value backwards-compatibility as much as in other languages, we may find out in a later python 3.x version that this will break entirely.

File: Makefile (hunk 1 above)
Comment (shereenElSayed) [reply Makefile]: I kept it this way to avoid making changes to the container.
@rdt12 Please confirm

File: Makefile (hunk 7 above)
Comment (shereenElSayed) [reply Makefile]: I will change it.

File: Makefile (hunk 2 above)
Comment (shereenElSayed) [reply Makefile]: I just aimed to keep it this way as well in case we do not need to make changes in the container

**Comment by shereenElSayed** [comment]
Thanks a lot, @myNameIsPatrick, I made the review comments, I just need to make sure that we can make changes to the container before making it as module+entrypoint only style and make the rest of your comments

**Review by myNameIsPatrick — APPROVED** [review]

**Pull request merged**